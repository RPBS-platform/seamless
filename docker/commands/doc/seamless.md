% seamless(1) Version 0.1 | seamless-cli Documentation

NAME
====

**seamless** - Starts a new Seamless Docker container and runs a command inside it.

SYNOPSIS
========

| **seamless** \[_command_]

DESCRIPTION
===========

Starts a new Seamless Docker container and runs a command inside it.

The current directory is mounted to /cwd, and the command is executed there

/tmp is mounted as well

Uses the host network for the Docker container. Will only work under Linux.

BUGS
====

See GitHub Issues: <https://github.com/sjdv1982/seamless/issues>

AUTHOR
======

Sjoerd de Vries <sjdv1982@gmail.com>

SEE ALSO
========

**seamless-bash(1)**, **seamless-ipython(1)**, **seamless-jupyter(1)**