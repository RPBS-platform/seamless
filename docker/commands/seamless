set -u -e

firstarg=$1 #there must be at least one argument
gdbflags='--cap-add=SYS_PTRACE --security-opt seccomp=unconfined --security-opt apparmor=unconfined'
docker run --rm \
  $gdbflags \
  --network=host \
  -v `pwd`:/cwd \
  -v /tmp:/tmp \
  --workdir /cwd \
  --user $(id -u):$(id -g) \
  rpbs/seamless $*
