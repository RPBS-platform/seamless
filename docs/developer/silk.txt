See plans/silk.md for a decent description of Silk.
Not everything in there has been implemented, but the essentials are there.