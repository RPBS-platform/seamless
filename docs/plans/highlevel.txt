Status of the high level:
- Most things are now working. Macro is still missing, node wrapping is very incomplete,
    and too easy to construct an illegal graph, but mostly fine now.
- hcell, hrc and htf refer to the node. Should be renamed.
- Translation is now lazy, i.e. only upon translate(), equilibrate()
- node["TEMP"] refers to a literal (i.e. Python) construction value. Low-level cells are needed to
  parse this to a serializable value. Right now, the mid-level does this on the first translation pass, then pops the TEMP.
  This is very ugly. mid-level should be configured to invoke remote parsing, and remote parsing should be invokable without
  translation. This way, the translate.py can be disconnected, and the remaining mid-and-high-level routines could run in
  the browser (under Pyodide; Transcrypt is probably less useful since we would want a Python REPL). If *all* of the state
  is in the graph (i.e. Python Context, Cell etc are just thin wrappers), then the graph could be manipulated in parallel
  by JS, e.g. manipulating the position of the node on a visualization canvas).
  Right now this is not exactly the case: there is context._graph but ALSO context._children.
