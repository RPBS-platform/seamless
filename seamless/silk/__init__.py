from .Silk import Silk, is_none
from jsonschema.exceptions import FormatError, ValidationError
