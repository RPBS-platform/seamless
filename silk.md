Silk schemas are a superset of JSON schema. A schema that validates against
a Silk schema will always validate against JSON schema too, if in plain form,
and if converted as below.

Seamless will implement its own $ref resolver for $refs that start with
SEAMLESS. These are provided by the central schema registrar.
These can be used for both full-blown schemas AND new basic types
(still refer to them using $ref instead of using {"type": ...} )

Conversion to JSON schema:
- SEAMLESS $refs are re-written
- constructs are replaced with their plain-form-generated schema

# extra schema fields
On top of JSON schema
(1) indicates object-only, (2) indicates array-only

## storage

### form
"plain": instance must be stored as a list, dictionary or plain scalar
"binary": instance is in binary form. This is a Numpy structured scalar for
objects, an Numpy array for arrays.
"any": no constraints on the form
Default: "any"

### (note on *form*)
The *actual* form of an instance is one of *four* values:
"pure-plain": plain, and does not contain binary (grand)children
"pure-binary": binary, and does not contain plain (grand)children
"mixed-plain": plain, but contains binary (grand)children
"mixed-binary":binary, but contains plain (grand)children
  (encoded as Python objects: see np.dtype.hasobject)

These *actual* forms exist only in the context of Silk wrapper instances,
not Silk-based schema validation.

Pure plain instances can be serialized as JSON, pure binary instances as Numpy arrays.
Mixed instances need special serialization.
For all schemas, Cython classes/structs can be generated. For any schema that is
not fixed-binary (see below), a Python object is filled in. (Fixed-length
  arrays of non-fixed-binary schemas are represented as (np.object, <length>) fields
  in the dtype; in Cython, this becomes a memoryview, for which size information
  is not as of yet supported)
Only for fixed-binary schemas, C struct headers can be generated.

Every Silk wrapper instance carries a descriptor of its own actual form
(serialized)
and a tuple of four counts, counting the number of forms for the children
(regenerated).
### /(note on *form*)

### dtype (only for scalars)
If form is binary, defines binary data type: float32, int32, uint8, etc.
Must match with "type": float32 for number, uint8 for integer, etc.

### gpu
Dict of storage parameters on the GPU.
Examples:
- OpenGL vertex buffer
- N-dimensional OpenGL texture
- OpenCL, CUDA
- Pointer to the GPU data (carried by Silk)

## /storage


## validators
See below.

## policy

For each policy field that is absent, the parent schema is queried.
If there is no parent schema, it is looked up in the global config.
(optimization: every Silk instance holds a ref to the nearest parent policy dict)
Default values are indicated for the Silk global config

Every infer_XXX property has a corresponding set_XXX function in the Silk API,
 which does the inference retroactively based on the current value(s).
Every set_XXX has a "recursive" parameter.

### surplus_to_binary
What to do with extra properties/items when plain is converted to binary
- Error
- Silently discard them
- Incorporate them as Python object fields (invalidates Cython/C headers generated from the schema)
Default: Error

### infer_property (2)
If a new property is accessed via Silk, an (empty) schema for that property is inserted under "properties"
If False, the returned Silk object will be schema-less (and policy-less)
For arrays, this has no meaning by itself, but still can be provided to govern the behavior of items
Default: True

### infer_type
Silk will replace a "any" or absent "type" value with the type of the first assigned value
For shaped arrays, this will write the type of the scalar base item in the schema
For non-shaped arrays, this has no meaning by itself, but still can be provided to govern the behavior of items
Default: True

### infer_dtype
Whenever an instance enters binary form (but not mixed-binary form),
the dtype is assigned based on the current value.
Default: Fakse

### infer_dtype_mixed
Same as above, but also for mixed-binary
Default: False

### force_valid_schema
Disallow manipulations of the schema that create an invalid schema, or that invalidate the data. If this happens,
the schema is error-buffered, if schema error buffering is enabled.
If force_valid_schema is disabled:
   If the schema is invalid, raise an exception.
   Else, the data is set to invalid, and schema violations are stored in the error log.
If there is neither an error log nor error buffering, force_valid_schema has no effect:
  an exception is raised, no matter what.
Default: True

### error_buffer (2)
Error-buffer the data.
Only works in plain form, disabled in binary form. Conversion to binary form will raise an exception if there is anything
in the error buffer.
Only works for "object" schemas. For other schemas, has no effect but can be used to control child behavior.
(error buffer can be cleared by API)
Any manipulation that is valid for the individual properties, but not for the instance as a whole,
is stored in a "diff" dict. What actually happens is as follows:
- After the individual property schemas have been successfully validated, a shallow copy of the old instance value is made
- The shallow copy is updated with the old "diff" dict (if any)
- The shallow copy is updated with the value of the newly manipulated property
- The shallow copy is validated.
  On success: the shallow copy replaces the instance value, and the "diff" dict (if any) is deleted.
  On failure: the old instance value is kept, and the newly manipulated property is added to the "diff" dict
In Silk, the (wrapper of the) manipulated property has access to its "raw value" from the "diff" dict.
If not enabled, an exception will be raised for illegal data manipulations
Default: True

### error_buffer_schema
Error-buffer all manipulations of the schema properties (see above).

## error_log
Errors (data-schema mismatches) are logged, rather than raising exceptions immediately.
Stack traces, exceptions and whatnot are stored in the error log.
In Silk, error logs are observable, both for appends and for clear events.
(error log can be cleared by API)
Default: True

### infer_item (1)
What happens if a new item is assigned:
"uniform": all items have the same schema. For the first item, an empty schema is created when it is first accessed via Silk.
"pluriform": all items have their own schema.  For each item, an empty schema is created when it is first accessed via Silk.
"none": If a new item is accessed via Silk, the returned Silk object will be schema-less (and policy-less)
Default: uniform

## infer_shape (1)
Whenever a Numpy array is assigned to an array, its shape parameters are imprinted on the array schema.
This turns the array schema into a fixed-shape schema.
The imprinting is done *after* the array is validated against the existing schema: so if the array was already fixed-shape,
re-imprinting is likely to fail.
For objects, this has no meaning by itself, but still can be provided to govern the behavior of properties
Default: False

## re_infer_shape (1)
Same as above, but existing shape-constraint schema parameters are first removed. This is useful if e.g. OpenCL/CUDA code needs to
be regenerated whenever the array shape changes.
Shape validators are not removed, so constraints on the shape can be encoded there.
Default: False

### infer_contents
New properties/items can be auto-created via Silk access
Default: True

### walk_tree
Arrays and objects are always tree structures. If a value is assigned to the array or object, walk the entire tree
of the assigned value and add a schema for each item.
Default: False

### binary_validation
If False:
  all schema validation is skipped, also for the children, if the instance is in binary form.
Default: True ( => not skipped)

### schema_undo_buffer
Allows you to undo schema manipulations (automatic and manual) that you didn't want to happen.
Works similar to the schema error buffer.
To-be-determined on what level the undo buffer will be stored (globally?), and the atomicity.

# Validators
Array of strings containing Python code. Each of them will be executed as an
eval'ed function.

Alternative (seamless world):
Array of "SEAMLESS:text.code.pythoncode". Each of them will be executed in a
transformer.

In any case, in its local namespace, the validator gets access to the object
in Silk form as "self", and to all properties as variables (same as for Spyder).
(Special case for construct validators: they get access to the construct instance descriptor instead)


# Fixed-binary schemas

Any instance can be converted to mixed-binary form. Many instances can be converted
to a pure-binary form as well: it is just that typically, the dtype/size in memory will depend
on the contents of the data.
A fixed-binary schema is a schema for which it is guaranteed that a pure-binary
form exists (no Python objects) and always has the same dtype/size in memory.
Only for fixed-binary schemas,
C struct headers can be generated. It is guaranteed
that the pure-binary form matches these struct headers.
(For binary conversion, Silk will generate Numpy dtypes with align=True).

Fixed-binary schemas are:
- non-string scalar schemas
- string schemas that have an upper length.
- object schemas whose properties are all fixed-binary
- array schemas with a fixed length and containing a single schema, which is
  itself a non-array fixed-binary schema.
- construct schemas that declare themselves fixed-binary  
This means that nested array schemas are *never* fixed-binary schemas.
shapedarrays can easily be fixed-binary, though.


# Constructs

Constructs are schemas where:
  "type" : "construct"
  "construct": <one of the registered constructs>

For Silk schema validation, every construct must register three functions:
1. From the "construct" schema dict, generate a schema for plain instances,
and another for binary/mixed-binary instances. If the construct schema is overall
invalid, this must be reported here (return None / exceptions).
2. From the instance, generate a construct instance descriptor.
This construct instance descriptor is validated using construct validators.
3. A function that returns if the binary schema is a pure-binary schema or not.

Note that a construct defines *extra* fields: the functions have access to all
other schema fields.

For Silk wrapping, a construct must also define the following:
1. A construct API is defined to manipulate instances of the construct.
For example, the shape construct API contains .append(), which does very different things
for plain or binary form instances.
2. A top-level converter between plain and binary form.

Normally, constructs also have a specialized generation routines for C struct headers
and Cython class headers generation.

# shapedarray construct
The shape construct is a straightforward construct around an N-dimensional Numpy array
or an equivalent nested-list plain representation. The Numpy array is space-efficient,
but not easy to grow or shrink.

## extra fields:

### shape
A list of items that are one of the following:
- None
- A single positive integer value
- A pair of (inclusive) lower-bound - upper-bound positive integer values
The length of the list must match with "ndim", if defined

### ndim
The array must have ndim dimensions. Must match with "shape", if defined.

### base
Schema of the scalar base item.
Base may be undefined, in which case the scalar base item is simply a Python
object.

### dtype
The binary storage data type (of the scalar base item) of the shaped array.

### c_contiguous
If form is binary, array must be c_contiguous
Default: False

### f_contiguous
If form is binary, array must be f_contiguous
Default: False

### validators
Construct validators for the shapedarray construct.
The shapedarray construct instance descriptor contains:
   .shape, .ndim, .dtype, .type (mapped from .dtype)
    .c_contiguous, .f_contiguous;
    in case of a Numpy array / Python buffer: .strides and some flags.

## /extra items

## Shapedarray pure-binary schema
shaped array schemas are pure-binary schemas if
- shaped array schemas h
- array schemas that are both *fully-shaped* (fixed-length in all dimensions)
   and whose items are all pure-binary schemas

# /shapedarray construct


# vararray construct
Construct that simulates a N-dimensional variable-length numpy array
using a fixed-length 1D "store" array. Offsets and sizes of each subarray
(or, in case of a 1D array, just the size of array itself) are stored. In case
of a simulated shape=(10,20,X) array, there are 200 subarray lengths,
and  the number of offsets is 201: the first offset
is 0, the last one is the length of the 1D store. Available space for each [x,y]
subarray can be computed as offset[(x,y) + 1] - offset[(x,y)] - length[(x,y)]
Accessing a subarray returns a wrapper around a Numpy array between the offsets.
Data access is restricted to :length[(x,y)], and asking for the raw Numpy array
also returns a Numpy array around store[offset[(x,y)]: offset[(x,y)] + length[(x,y)]]
API supports .append() etc. which is rather efficient as long as there is there is space.
Policy options to define the initial amount of space upon initialization/to-binary conversion,
 and what happens if there is an append() and no space.

Multiple variable-length dimensions are also possible: shape(10, X, Y), but lengths
must be stored independently per dimension (the X'es and the Y's).
.append() to the second dimension (increasing one of the 10 X'es) is again rather
efficient, but it must be policy'ed how much space a newly created Y receives.

vararray schemas are pure-binary schemas as long as the base scalar item is a
pure-binary schema.

# tree construct
The tree consists of nodes and leaves. Nodes contain other nodes, and leaves.
Nodes are stored as a vararray of indices, where every index stands for a node or for a leaf,
depending on its value (i.e. negative is node, positive is leaf).
Leaves are stored as a 1D array, into which the nodes index.
Policy: store the parent node index for every node (or 0 for the top node).

tree schemas are pure-binary schemas as long as the node schema is a
pure-binary schema.

# other possible constructs
(typed) sets, (typed) maps/hash tables, linked lists, ...
"Internal" lookup-accelerating structures etc. can be stored as well, by policy,
 as long as they use indices and not raw pointers.
