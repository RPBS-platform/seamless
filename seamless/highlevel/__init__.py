import inspect
from copy import deepcopy
from types import LambdaType
from ast import PyCF_ONLY_AST, FunctionDef, Expr, Lambda

from ..mixed import MixedBase
from ..silk import Silk
from ..silk.validation import _allowed_types
from ..core.lambdacode import lambdacode
from ..core.cached_compile import cached_compile

ConstantTypes = _allowed_types + (Silk, MixedBase, tuple)

def set_hcell(cell, value):
    from ..core.structured_cell import StructuredCellState
    if cell["celltype"] == "structured":
        cell["stored_state"] = StructuredCellState.from_data(value)
    else:
        cell["stored_value"] = value

def parse_function_code(code_or_func, identifier="<None>"):
    if callable(code_or_func):
        func = code_or_func
        code = inspect.getsource(func)
        if isinstance(func, LambdaType) and func.__name__ == "<lambda>":
            code = lambdacode(func)
            if code is None:
                raise ValueError("Cannot extract source code from this lambda")
    else:
        assert isinstance(code_or_func, str)
        code = code_or_func

    ast = cached_compile(code, identifier, "exec", PyCF_ONLY_AST)
    is_function = (len(ast.body) == 1 and
                   isinstance(ast.body[0], FunctionDef))

    if is_function:
        func_name = ast.body[0].name
        code_object = cached_compile(code, identifier, "exec")
    else:
        assert (len(ast.body) == 1 and isinstance(ast.body[0], Expr))
        assert isinstance(ast.body[0].value, Lambda)
        func_name = "<lambda>"
        code_object = cached_compile(code, identifier, "eval")
    return code, func_name, code_object

from .Context import Context
from .Library import stdlib, mylib
from .Reactor import Reactor
from .Transformer import Transformer
from .Cell import Cell
from .Link import Link
