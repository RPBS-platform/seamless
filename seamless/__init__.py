"""
Seamless: framework for data-driven and live programming
Copyright 2016-2018, Sjoerd de Vries
"""

import sys
import time
import functools
import traceback
import atexit

from abc import abstractmethod
class Wrapper:
    @abstractmethod
    def _unwrap(self):
        pass

from . import silk
from . import mixed

#Dependencies of seamless

# 1. hard dependencies; without these, "import seamless" will fail.
# Still, if necessary, some of these dependencies could be removed, but seamless would have to be more minimalist in loading its lib

import numpy as np

if np.dtype(np.object).itemsize != 8:
    raise ImportError("Seamless requires a 64-bit system")

#silk must be imported before mixed
from . import silk
from . import mixed

from .core import mainloop as _mainloop
from .core.mainloop import mainloop, mainloop_one_iteration, asyncio_finish, workqueue
atexit.register(asyncio_finish)

ipython = None
ipy_error = None
try:
    from IPython import get_ipython
    from IPython.core.error import UsageError
    from IPython.terminal.pt_inputhooks import register as _register_integration_terminal
    from ipykernel.eventloops import register_integration as _register_integration_kernel
except ImportError:
    raise
    ipy_error = "Cannot find IPython"
else:
    ipython = get_ipython()
    if ipython is None:
        ipy_error = "Seamless was not imported inside IPython"

if ipy_error is None:
    last_exception = None
    def new_except_hook(etype, evalue, tb):
        global last_exception
        exc = traceback.format_exception(etype, evalue, tb)
        if exc != last_exception:
            last_exception = exc
            print("".join(exc))

    def patch_excepthook():
        sys.excepthook = new_except_hook
    patch_excepthook()

    def mainloop():
        raise RuntimeError("Cannot run seamless.mainloop() in IPython mode")

    def inputhook_terminal(context):
        while not context.input_is_ready():
            mainloop_one_iteration()
    _register_integration_terminal("seamless", inputhook_terminal)

    @_register_integration_kernel('seamless')
    def inputhook_kernel(kernel):
        while 1:
            t = time.time()
            while time.time() - t < kernel._poll_interval:
                mainloop_one_iteration()        
            kernel.do_one_iteration()

    _register_integration_terminal("seamless", inputhook_terminal)
    
    ipython.enable_gui("seamless")
    
else:
    sys.stderr.write("    " + ipy_error + "\n")
    sys.stderr.write("    Call seamless.mainloop(), seamless.flush() or context.equilibrate() to process cell updates\n")

def flush():
    from .core.mainloop import workqueue
    workqueue.flush()

from .highlevel import *
from .silk import Silk
from .debugger import pdb
