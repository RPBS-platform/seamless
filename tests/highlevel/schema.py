'''
from seamless.silk import Silk

buf = {}
s = Silk(buffer=buf)
s.x = 20
s.y = 88
def validate(self):
    print("VALIDATE", self)
    assert self.x > 0
s.add_validator(validate)
s.x = -1
print(s)
print(s.data)
print(buf)
'''

from seamless.highlevel import Context

ctx = Context()
ctx.mount_graph("/tmp/seamless", persistent=True)
ctx.a = {}
ctx.a.x = 20
ctx.a.y = 88
def validate(self):
    print("VALIDATE", self)
    assert self.x > 0
ctx.a.handle.add_validator(validate)
ctx.a.x = -1
print("OK")
print(ctx.a._get_cell().value)
print(ctx.a._get_cell().handle)
