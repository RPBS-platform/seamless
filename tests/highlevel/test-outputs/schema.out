VALIDATE <Silk: {'x': -1, 'y': 88} >
OK
<Silk: None >
<Silk: {'x': -1, 'y': 88} >
